### PR TITLE
Add automated migration for Destinations to Notification configs

### PR DESCRIPTION
**Pull request by qreshi**
*Issue #, if available:* #105

*Description of changes:*
This PR adds a background job that will migrate Destinations to Notification configs as part of Alerting's integration with the Notification plugin for a more seamless transition to 2.0.

* The background job will be triggered on a cluster change event (if we feel the need to have further control on this, we can revisit adding a plugin setting to re-enable or permanently disable this job as well)
* The background job once scheduled will run every minute and attempt to move all Destinations to Notification configs, once all Destinations are moved over, the job will disable itself
* The migration job has a lock that is used to ensure that there is no more than one instance of the job running at a time
* Once a Destination is migrated over to a Notification config, the Destination will be deleted
* The migrated Notification configs will be created with the same ID as the Destinations so that the Monitor doesn't have to have any changes to its references in the Action definition (in a later PR, the MonitorRunner will be updated to be able to execute Actions seamlessly regardless of whether or not the underlying config has been migrated)

*CheckList:*
[x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/alerting/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by aditjind** [comment]
> The MonitorRunner will be updated to be able to execute Actions seamlessly regardless of whether or not the underlying config has been migrated. 

Does this mean `MonitorRunner` will have the support to execute actions by reading configs from both the `NotificationConfig` and the `AlertingConfig`? Any specific reason we need to support both since we already have a job with features such as locking support etc.

File: alerting/src/test/kotlin/org/opensearch/alerting/util/destinationmigration/DestinationMigrationUtilServiceIT.kt [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+                      "destination" : ${destination.toJsonString()}
+                    }
+                """.trimIndent()
+                indexDoc(SCHEDULED_JOBS_INDEX, destination.id, dest)
+                ids.add(destination.id)
+            }
+
+            // Create cluster change event and wait for migration service to complete migrating data over
+            client().updateSettings("indices.recovery.max_bytes_per_sec", "40mb")
+            Thread.sleep(120000)
```
Comment (lezzago) [inline alerting/src/test/kotlin/org/opensearch/alerting/util/destinationmigration/DestinationMigrationUtilServiceIT.kt:84]: Is there a way to reduce this time without making it flaky?

File: alerting/src/main/kotlin/org/opensearch/alerting/util/destinationmigration/DestinationMigrationUtilService.kt [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    companion object {
+
+        private val logger = LogManager.getLogger(DestinationMigrationUtilService::class)
+
+        @Volatile
+        private var runningLock = false // In case 2 migrateDestinations() processes are running
+
+        // Used in DestinationMigrationCoordinator to cancel scheduled process
+        @Volatile
+        var finishFlag = false
```
Comment (lezzago) [inline alerting/src/main/kotlin/org/opensearch/alerting/util/destinationmigration/DestinationMigrationUtilService.kt:53]: Did we want to make this flag a cluster setting to get reset later on?

File: alerting/src/main/kotlin/org/opensearch/alerting/util/destinationmigration/DestinationMigrationUtilService.kt [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+            }
+            try {
+                runningLock = true
+
+                val destinationsToMigrate = retrieveDestinationsToMigrate(client)
+                logger.info("Need to migrate ${destinationsToMigrate.size} destinations")
+                if (destinationsToMigrate.isEmpty()) {
+                    finishFlag = true
+                    runningLock = false
+                    return
```
Comment (lezzago) [inline alerting/src/main/kotlin/org/opensearch/alerting/util/destinationmigration/DestinationMigrationUtilService.kt:72]: Need to delete the old destinations after the migration still

File: alerting/src/main/kotlin/org/opensearch/alerting/util/destinationmigration/DestinationMigrationUtilService.kt (hunk 2 above)
Comment (qreshi) [reply alerting/src/main/kotlin/org/opensearch/alerting/util/destinationmigration/DestinationMigrationUtilService.kt:53]: Let's treat that as a potential enhancement for now. If we see a need for it, we can follow-up to add it.

File: alerting/src/test/kotlin/org/opensearch/alerting/util/destinationmigration/DestinationMigrationUtilServiceIT.kt (hunk 1 above)
Comment (qreshi) [reply alerting/src/test/kotlin/org/opensearch/alerting/util/destinationmigration/DestinationMigrationUtilServiceIT.kt:84]: Hmm, the job will wait 1 minute before actually running for the first time. So if we think once running it doesn't take too long, we might be able to reduce it to as little over 1 minute as possible.

I won't treat it as a blocker for merging if that's fine with you but we can follow-up on it during general testing.

**Comment by qreshi** [comment]
> > The MonitorRunner will be updated to execute Actions seamlessly whether we have migrated the underlying config.
> 
> Does this mean `MonitorRunner` will have the support to execute actions by reading configs from both the `NotificationConfig` and the `AlertingConfig`? Any specific reason we need to support both since we already have a job with features such as locking support etc.

That's correct, the MonitorRunner will have changes as a follow-up to this PR so that the Monitor can continue to run at any time regardless of which Destinations has been migrated or not.

The `runningLock` for `DestinationMigrationCoordinator` is not going to halt Monitor executions during migration, that's just to ensure that multiple instances of the migration job aren't running concurrently.

File: alerting/src/test/kotlin/org/opensearch/alerting/util/destinationmigration/DestinationMigrationUtilServiceIT.kt (hunk 1 above)
Comment (lezzago) [reply alerting/src/test/kotlin/org/opensearch/alerting/util/destinationmigration/DestinationMigrationUtilServiceIT.kt:84]: sure thats fine

File: alerting/src/main/kotlin/org/opensearch/alerting/util/destinationmigration/NotificationApiUtils.kt [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+import org.opensearch.commons.notifications.action.DeleteNotificationConfigRequest
+import org.opensearch.commons.notifications.action.DeleteNotificationConfigResponse
+import org.opensearch.commons.notifications.action.GetNotificationConfigRequest
+import org.opensearch.commons.notifications.action.GetNotificationConfigResponse
+import org.opensearch.commons.notifications.action.SendNotificationRequest
+import org.opensearch.commons.notifications.action.SendNotificationResponse
+import org.opensearch.commons.notifications.action.UpdateNotificationConfigRequest
+import org.opensearch.commons.notifications.action.UpdateNotificationConfigResponse
+
+class NotificationApiUtils {
```
[the remark this answers is not in the document]
Comment (lezzago) [reply alerting/src/main/kotlin/org/opensearch/alerting/util/destinationmigration/NotificationApiUtils.kt:27]: There will only be a single implementation on Alerting for this and this is needed to talk to Notification, so I dont see any benefit with abstracting it

**Review by lezzago — APPROVED** [review]

File: alerting/src/main/kotlin/org/opensearch/alerting/util/destinationmigration/NotificationApiUtils.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (aditjind) [reply alerting/src/main/kotlin/org/opensearch/alerting/util/destinationmigration/NotificationApiUtils.kt:27]: Cleaner code maybe? I mean we can keep it as it is for now.

[the remark this answers is not in the document]
Comment (qreshi) [reply alerting/src/main/kotlin/org/opensearch/alerting/util/destinationmigration/NotificationApiUtils.kt:27]: Yeah, like Ashish said, there won't be multiple implementations of this so an interface wouldn't be too helpful for the utils.

**Review by AWSHurneyt — APPROVED** [review]

**Pull request merged**